plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/ExecutionStats.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.exchange;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.stats.TimeStat;
+import org.weakref.jmx.Managed;
+import org.weakref.jmx.Nested;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
+
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class ExecutionStats
+{
+    private final TimeStat finished = new TimeStat(MILLISECONDS);
+    private final TimeStat failed = new TimeStat(MILLISECONDS);
+
+    public <T> T record(Supplier<T> call)
+    {
+        long start = System.currentTimeMillis();
+        try {
+            T result = call.get();
+            long finish = System.currentTimeMillis();
+            finished.add(finish - start, MILLISECONDS);
+            return result;
+        }
+        catch (Throwable t) {
+            long finish = System.currentTimeMillis();
+            failed.add(finish - start, MILLISECONDS);
+            throw t;
+        }
+    }
+
+    public <T> CompletableFuture<T> record(CompletableFuture<T> future)
+    {
+        long start = System.currentTimeMillis();
+        future.whenComplete((value, failure) -> {
+            long finish = System.currentTimeMillis();
+            if (failure == null) {
+                finished.add(finish - start, MILLISECONDS);
+            }
+            else {
+                failed.add(finish - start, MILLISECONDS);
+            }
+        });
+        return future;
+    }
+
+    public <T> ListenableFuture<T> record(ListenableFuture<T> future)
+    {
+        long start = System.currentTimeMillis();
+        Futures.addCallback(future, new FutureCallback<T>()
+        {
+            @Override
+            public void onSuccess(T result)
+            {
+                finished.add(System.currentTimeMillis() - start, MILLISECONDS);
+            }
+
+            @Override
+            public void onFailure(Throwable t)
+            {
+                failed.add(System.currentTimeMillis() - start, MILLISECONDS);
+            }
+        }, directExecutor());
+        return future;
+    }
+
+    @Managed
+    @Nested
```

Comment:
For my understanding, can you briefly explain what these two annotations do?

Reply:
Methods annotated with `@Managed` are exported via JMX. `@Nested` tells the framework to recurse into the object returned by a method and export nested methods annotated with `@Managed`.